### PR TITLE
HHH-13614 - Allow the IntegratorProvider to be supplied via its FQN in the JPA persistence.xml

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-13614

### CODE REVIEW

**Comment by dreab8** [comment]
hi @vladmihalcea the PR applies few changes to your https://github.com/hibernate/hibernate-orm/pull/3032

File: hibernate-core/src/main/java/org/hibernate/jpa/boot/internal/EntityManagerFactoryBuilderImpl.java [hunk 1]
```diff
@@ -1394,4 +1382,51 @@ private void addCacheRegionDefinition(CacheRegionDefinition cacheRegionDefinitio
 			this.cacheRegionDefinitions.add( cacheRegionDefinition );
 		}
 	}
+
+	private <T> T loadSettingInstance(String settingName, Object settingValue, Class<T> clazz) {
+		T instance = null;
+		Class<? extends T> instanceClass = null;
+
+		if ( clazz.isInstance( settingValue ) ) {
```
Comment (dreab8) [inline hibernate-core/src/main/java/org/hibernate/jpa/boot/internal/EntityManagerFactoryBuilderImpl.java]: not sure if this should be `if ( clazz.isInstance( settingValue ) ) `or `if ( clazz.isAssignableFrom( settingValue.getClass() ) )`

Comment (vladmihalcea) [reply hibernate-core/src/main/java/org/hibernate/jpa/boot/internal/EntityManagerFactoryBuilderImpl.java]: The `isInstance` check allows us to pass an actual `IntegratorProvider` instance that was build programatically. We could support both actual instances and `Class<? extend IntegratorProvider>` that allow us to instantiate the `IntegratorProvider`.

**Comment by vladmihalcea** [comment]
@dreab8 Looks good. Much better to split the logic into dedicated methods.

**Comment by Sanne** [comment]
I'm failing to see the use case. Integrator Providers can be automatically picked up from classpath, the reason is takes an object instance is to allow for this to  work in modular / managed environments - but that wouldn't work with a FQN as you can't set the classloader.

So what's the point to have this?

**Comment by vladmihalcea** [comment]
For Spring Boot, you could define the IntegratorProvider in the application.properties file. Otherwise, you need to do all sorts of hacks to pass an instance.

**Comment by Sanne** [comment]
ok, thanks @vladmihalcea . Weird that they don't use the automatic integration facility but I guess that's their choice.

**Comment by vladmihalcea** [comment]
@Sanne I wrote [several articles](https://vladmihalcea.com/hibernate-event-listeners/) that show the advantages of using the `IntegratorProvider`, and several people commented that they can't use the functionality easily with Spring Boot.

Spring Boot support

That's when I noticed the opportunity for having this enhancement.

File: hibernate-core/src/main/java/org/hibernate/jpa/boot/internal/EntityManagerFactoryBuilderImpl.java (hunk 1 above)
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/jpa/boot/internal/EntityManagerFactoryBuilderImpl.java]: @vladmihalcea reverted to` clazz.isAssignableFrom( settingValue.getClass())`

**Comment by dreab8** [comment]
added few more tests, rebased and merged upstream.

Thanks a lot @vladmihalcea

**Comment by vladmihalcea** [comment]
@dreab8 Cool! I'm going to update my articles once this will be released.

**Comment by Sanne** [comment]
Nice, thanks @vladmihalcea . Keep in mind that if you declare your Integrator as a manifest, ORM can pick it up without any configuration being needed. For example Hibernate Validator and Hibernate Search are automatically integrated - no need to configure them, one just needs to add the dependency.

**Pull request closed without merging**